File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.cpp
// Last updated: 17/07/2025, 20:51:38
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void f(TreeNode* root, int& count, int maxi){
        if(root == NULL){
            return  ;
        }
        if(root->val >= maxi){
            count++;
            maxi = root->val;
        }
        f(root->left, count, maxi);
        f(root->right, count, maxi);
    }
    int goodNodes(TreeNode* root) {
        int count = 0;
        int maxi = INT_MIN;
        f(root, count, maxi);
        return count;
    }
};

File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.py
# Last updated: 01/10/2025, 00:12:21
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        def dfs(node, max_so_far):
            if not node:
                return 0
            
            good = 1 if node.val >= max_so_far else 0

            new_max = max(max_so_far, node.val)
            left_good = dfs(node.left, new_max)
            right_good = dfs(node.right, new_max)

            return good + left_good+ right_good
        
        return dfs(root, root.val)
        